provider/src/test/kotlin/org/gradle/kotlin/dsl/support/ImplicitImportsTest.kt

```diff
@@ -0,0 +1,42 @@
+package org.gradle.kotlin.dsl.support
+
+import org.gradle.kotlin.dsl.fixtures.AbstractIntegrationTest
+import org.hamcrest.CoreMatchers.equalTo
+import org.junit.Assert.assertThat
+import org.junit.Test
+import java.io.File
+import java.util.jar.JarFile
+
+class ImplicitImportsTest : AbstractIntegrationTest() {
+
+    @Test
+    fun `implicit imports are fully qualified to allow use of the preferred type amongst those with same simple name in different Gradle API packages`() {
+
+        // given:
+        existing("settings.gradle").appendText("""
+            rootProject.name = "some-project"
+        """)
+        withBuildScript("""
+
+            plugins {
+                java
+            }
+
+            // Prefer org.gradle.api.tasks.bundling.Jar over org.gradle.jvm.tasks.Jar
+            tasks.withType<Jar> {
```

Comment:
Do you want to add something like `assert(this is org.gradle.api.tasks.bundling.Jar)` just to be explicit?

Also, what is the difference between `org.gradle.api.tasks.bundling.Jar` and `org.gradle.jvm.tasks.Jar`?

Reply:
> Also, what is the difference between org.gradle.api.tasks.bundling.Jar and org.gradle.jvm.tasks.Jar?

[`org.gradle.jvm.tasks.Jar`](https://github.com/gradle/gradle/blob/685d9df6b2437908c49ba6e2fd54ed8c6c7ff0ac/subprojects/platform-jvm/src/main/java/org/gradle/jvm/tasks/Jar.java) is the real task

[org.gradle.api.tasks.bundling.Jar](https://github.com/gradle/gradle/blob/a3c1bd72dc1fbee660b6b7da8cbf57e1758c83ac/subprojects/plugins/src/main/java/org/gradle/api/tasks/bundling/Jar.java) is a remnant of past refactoring, here for backwards compatibility

I mixed up the two in the comments that can be seen the line above your comment, sorry about the confusion.